src/coordination/coordinator_client.cpp

```diff
@@ -27,82 +27,78 @@ auto CreateClientContext(const memgraph::coordination::CoordinatorClientConfig &
 }
 }  // namespace
 
-CoordinatorClient::CoordinatorClient(const CoordinatorClientConfig &config)
+CoordinatorClient::CoordinatorClient(CoordinatorData *coord_data, CoordinatorClientConfig config,
+                                     HealthCheckCallback succ_cb, HealthCheckCallback fail_cb)
     : rpc_context_{CreateClientContext(config)},
       rpc_client_{io::network::Endpoint(io::network::Endpoint::needs_resolving, config.ip_address, config.port),
                   &rpc_context_},
-      config_{config} {}
-
-CoordinatorClient::~CoordinatorClient() {
-  auto exit_job = utils::OnScopeExit([&] {
-    StopFrequentCheck();
-    thread_pool_.Shutdown();
-  });
-  const auto endpoint = rpc_client_.Endpoint();
-  // Logging can throw
-  spdlog::trace("Closing replication client on {}:{}", endpoint.address, endpoint.port);
-}
+      config_{std::move(config)},
+      coord_data_{coord_data},
+      succ_cb_{std::move(succ_cb)},
+      fail_cb_{std::move(fail_cb)} {}
+
+auto CoordinatorClient::InstanceName() const -> std::string { return config_.instance_name; }
+auto CoordinatorClient::SocketAddress() const -> std::string { return rpc_client_.Endpoint().SocketAddress(); }
 
 void CoordinatorClient::StartFrequentCheck() {
   MG_ASSERT(config_.health_check_frequency_sec > std::chrono::seconds(0),
             "Health check frequency must be greater than 0");
-  replica_checker_.Run(
-      "Coord checker", config_.health_check_frequency_sec,
-      [last_response_time = &last_response_time_, rpc_client = &rpc_client_] {
+
+  instance_checker_.Run(
+      "Coord checker", config_.health_check_frequency_sec, [this, instance_name = config_.instance_name] {
         try {
-          {
-            auto stream{rpc_client->Stream<memgraph::replication_coordination_glue::FrequentHeartbeatRpc>()};
-            stream.AwaitResponse();
-            last_response_time->store(std::chrono::system_clock::now(), std::memory_order_acq_rel);
-          }
+          spdlog::trace("Sending frequent heartbeat to machine {} on {}", instance_name,
+                        rpc_client_.Endpoint().SocketAddress());
+          auto stream{rpc_client_.Stream<memgraph::replication_coordination_glue::FrequentHeartbeatRpc>()};
+          stream.AwaitResponse();
+          succ_cb_(coord_data_, instance_name);
         } catch (const rpc::RpcFailedException &) {
-          // Nothing to do...wait for a reconnect
+          fail_cb_(coord_data_, instance_name);
```

Comment:
Maybe a todo:
1) Try to avoid searching for the instance by name.
2) Have a single thread that goes through all instances.

Reply:
1. there will be at most few instances, that's why searching them is basically constant operation in time
2. Currently, only one thread is searching instances, the one which is executing the code, could you explain a bit more?